pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -107,20 +114,47 @@ func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Ba
 	})
 }
 
-func getPrivateIPAddress(nic *network.Interface) (string, error) {
+func getPrivateIPv4Address(nic *network.Interface) (string, error) {
+	if nic.IPConfigurations == nil {
+		return "", fmt.Errorf("nic.IPConfigurations %s is nil", *nic.ID)
+	}
+
+	ipConfigurations := *nic.IPConfigurations
+	for _, ipConfiguration := range ipConfigurations {
+		if ipConfiguration.PrivateIPAddress != nil {
+			ipv4 := net.ParseIP(*ipConfiguration.PrivateIPAddress).To4()
+			if ipv4 != nil {
+				return ipv4.String(), nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("failed to get IPv4 PrivateIPAddress on nic %s", *nic.ID)
+}
+
+func getPrivateIPv6Address(nic *network.Interface) (string, error) {
 	if nic.IPConfigurations == nil {
 		return "", fmt.Errorf("nic.IPConfigurations %s is nil", *nic.ID)
 	}
 
 	ipConfigurations := *nic.IPConfigurations
-	if len(ipConfigurations) > 0 && ipConfigurations[0].PrivateIPAddress != nil {
-		return *ipConfigurations[0].PrivateIPAddress, nil
+	for _, ipConfiguration := range ipConfigurations {
+		if ipConfiguration.PrivateIPAddress != nil {
+			ip := net.ParseIP(*ipConfiguration.PrivateIPAddress)
+			if len(ip.To4()) == net.IPv4len {
+				continue
+			}
```

Comment:
why don't you just check if `ip.To4() == nil` to figure out if it's an IPv4 address? From the documentation it says
> If ip is not an IPv4 address, To4 returns nil.

from the implementation you can also see that there is a case where  len(ip) == IPv6len but it's still an IPv4 address.
```
func (ip IP) To4() IP {
	if len(ip) == IPv4len {
		return ip
	}
	if len(ip) == IPv6len &&
		isZeros(ip[0:10]) &&
		ip[10] == 0xff &&
		ip[11] == 0xff {
		return ip[12:16]
	}
	return nil
}
```

Reply:
Hi,
`len(ip) == IPv6len but it's still an IPv4 address` it's not... because of `return ip[12:16]`.

I checked the code once more but feels like it's slightly better to keep it like this to keep the code more consistent because `ip.To16()` implementation can return nil or IPv4 and check for length is the only way...
```
func (ip IP) To16() IP {
	if len(ip) == IPv4len {
		return IPv4(ip[0], ip[1], ip[2], ip[3])
	}
	if len(ip) == IPv6len {
		return ip
	}
	return nil
}
```

Reply:
yes you are right, I had the wrong assumption but as you are comparing the length of the To4() result and not of `ip` it should be fine.